{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook to run differential expression analysis using ´MAST´\n",
    "\n",
    "**Created by** : **Srivalli Kolla**\n",
    "\n",
    "**Devloped on** : July 19,2024 \n",
    "\n",
    "**Last Modified** : July 19,2024 \n",
    "\n",
    "**Institute of Systems Immunology, University of Wurzburg**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Importing packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Installed version </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.12</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">1</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> of sc-toolbox is outdated. Newest version is </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.12</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">3</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">!</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mInstalled version \u001b[0m\u001b[1;31m0.12\u001b[0m\u001b[1;31m.\u001b[0m\u001b[1;31m1\u001b[0m\u001b[1;31m of sc-toolbox is outdated. Newest version is \u001b[0m\u001b[1;31m0.12\u001b[0m\u001b[1;31m.\u001b[0m\u001b[1;31m3\u001b[0m\u001b[1;31m!\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "import glob\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import sc_toolbox\n",
    "import scipy.io\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.sparse as sparse\n",
    "from rpy2 import robjects\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects.packages import importr\n",
    "\n",
    "pandas2ri.activate()\n",
    "\n",
    "MAST = importr('MAST')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.8.0\n",
      "scanpy      1.9.3\n",
      "-----\n",
      "PIL                 8.2.0\n",
      "adjustText          1.2.0\n",
      "argcomplete         NA\n",
      "attr                23.2.0\n",
      "backcall            0.2.0\n",
      "backports           NA\n",
      "backports_abc       NA\n",
      "beta_ufunc          NA\n",
      "binom_ufunc         NA\n",
      "cffi                1.15.1\n",
      "cloudpickle         2.2.1\n",
      "colorama            0.4.6\n",
      "cycler              0.10.0\n",
      "cython_runtime      NA\n",
      "dask                2022.02.0\n",
      "dateutil            2.9.0\n",
      "debugpy             1.6.3\n",
      "decorator           5.1.1\n",
      "entrypoints         0.4\n",
      "fsspec              2023.1.0\n",
      "h5py                3.7.0\n",
      "ipykernel           6.16.2\n",
      "ipython_genutils    0.2.0\n",
      "jedi                0.19.1\n",
      "jinja2              3.1.4\n",
      "joblib              1.3.2\n",
      "kiwisolver          1.4.4\n",
      "llvmlite            0.39.1\n",
      "markupsafe          2.1.5\n",
      "matplotlib          3.5.3\n",
      "more_itertools      NA\n",
      "mpl_toolkits        NA\n",
      "natsort             8.4.0\n",
      "nbinom_ufunc        NA\n",
      "numba               0.56.3\n",
      "numpy               1.21.6\n",
      "packaging           23.2\n",
      "pandas              1.3.5\n",
      "parso               0.8.4\n",
      "pexpect             4.9.0\n",
      "pickleshare         0.7.5\n",
      "pkg_resources       NA\n",
      "prompt_toolkit      3.0.47\n",
      "psutil              5.9.3\n",
      "ptyprocess          0.7.0\n",
      "pydev_ipython       NA\n",
      "pydevconsole        NA\n",
      "pydevd              2.8.0\n",
      "pydevd_file_utils   NA\n",
      "pydevd_plugins      NA\n",
      "pydevd_tracing      NA\n",
      "pygments            2.17.2\n",
      "pyparsing           3.1.2\n",
      "pypi_latest         0.1.2\n",
      "pytz                2024.1\n",
      "questionary         1.10.0\n",
      "rich                NA\n",
      "rpy2                3.5.16\n",
      "sc_toolbox          0.12.1\n",
      "scipy               1.7.3\n",
      "seaborn             0.12.2\n",
      "session_info        1.0.0\n",
      "six                 1.16.0\n",
      "sklearn             1.0.2\n",
      "sparse              0.13.0\n",
      "statsmodels         0.13.2\n",
      "storemagic          NA\n",
      "threadpoolctl       3.1.0\n",
      "tlz                 0.12.1\n",
      "toolz               0.12.1\n",
      "tornado             6.2\n",
      "traitlets           5.9.0\n",
      "typing_extensions   NA\n",
      "tzlocal             NA\n",
      "wcwidth             0.2.10\n",
      "yaml                6.0\n",
      "zipp                NA\n",
      "zmq                 24.0.1\n",
      "-----\n",
      "IPython             7.33.0\n",
      "jupyter_client      7.4.9\n",
      "jupyter_core        4.11.1\n",
      "notebook            6.5.7\n",
      "-----\n",
      "Python 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 06:08:21) [GCC 9.4.0]\n",
      "Linux-6.9.7-100.fc39.x86_64-x86_64-with-fedora-39-Thirty_Nine\n",
      "-----\n",
      "Session information updated at 2024-07-22 17:44\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 3\n",
    "sc.logging.print_versions()\n",
    "sc.settings.set_figure_params(dpi = 300, color_map = 'RdPu', dpi_save = 300, vector_friendly = True, format = 'svg')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MAST"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Preparation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Steps followed\n",
    "\n",
    "1. Data loading and subsetting\n",
    "2. Normalization and log transformation\n",
    "3. Generating individual files for genes, barcodes, metadata and a matrix having log counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filtered out 1773 genes that are detected in less than 10 counts\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/skolla/miniforge3/envs/mast_env/lib/python3.7/site-packages/scanpy/preprocessing/_simple.py:249: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata.var['n_counts'] = number\n",
      "/home/skolla/miniforge3/envs/mast_env/lib/python3.7/contextlib.py:119: FutureWarning: X.dtype being converted to np.float32 from float64. In the next version of anndata (0.9) conversion will not be automatic. Pass dtype explicitly to avoid this warning. Pass `AnnData(X, dtype=X.dtype, ...)` to get the future behavour.\n",
      "  next(self.gen)\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read_h5ad('/home/skolla/Github/hofmann_dmd/hofmann_dmd/data/concatenated_CMC-16_07_2024.h5ad')\n",
    "\n",
    "desired_cell_states = ['vCM1', 'vCM2', 'vCM3', 'vCM4']\n",
    "adata = adata[adata.obs['cell_states'].isin(desired_cell_states)]\n",
    "\n",
    "\n",
    "sc.pp.filter_genes(adata, min_counts=10)\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "\n",
    "\n",
    "sparse_matrix = sparse.csr_matrix(adata.X.T)\n",
    "scipy.io.mmwrite('/home/skolla/Github/hofmann_dmd/hofmann_dmd/DE/de_results/concat_matrix.mtx', sparse_matrix)\n",
    "genes = pd.DataFrame(adata.var.index, columns=[\"gene\"])\n",
    "genes.to_csv('/home/skolla/Github/hofmann_dmd/hofmann_dmd/DE/de_results/concat_genes.tsv', sep='\\t', index=False, header=False)\n",
    "barcodes = pd.DataFrame(adata.obs.index, columns=[\"barcode\"])\n",
    "barcodes.to_csv('/home/skolla/Github/hofmann_dmd/hofmann_dmd/DE/de_results/concat_barcodes.tsv', sep='\\t', index=False, header=False)\n",
    "metadata = adata.obs\n",
    "metadata.to_csv('/home/skolla/Github/hofmann_dmd/hofmann_dmd/DE/de_results/concat_metadata.tsv', sep='\\t', index=True, header=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defining Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Steps followed in defining functions\n",
    "\n",
    "*Create_single_cell_assay*\n",
    "\n",
    "For creating a single cell assay object which can be used by MAST for differential gene expression\n",
    "\n",
    "1. Read the matrix, genes, and barcodes files\n",
    "2. Ensure the matrix has correct row and column names by checking with colnames and riownames of matrix\n",
    "3. Convert sparse matrix to dense matrix\n",
    "4. Ensure metadata matches the expression matrix\n",
    "5. Create the SingleCellAssay object\n",
    "6. Compute number of expressed genes per cell and store as a column in sca object\n",
    "7. Store columns of interest as factors\n",
    "8. Create the SingleCellAssay object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to subset the data to only include the target cell state and genotype\n",
    "def subset_data(adata, target_cell_state, target_genotype, groupby_cell=\"cell_states\", groupby_genotype=\"genotype\"):\n",
    "    # Subset the AnnData object to only include cells from the target cell state and genotype\n",
    "    adata_subset = adata[(adata.obs[groupby_cell] == target_cell_state) & (adata.obs[groupby_genotype] == target_genotype)].copy()\n",
    "    return adata_subset\n",
    "\n",
    "\n",
    "# Function to convert the AnnData subset to SingleCellAssay object (sca) for MAST analysis\n",
    "def convert_to_sca(adata_subset):\n",
    "    print(\"Converting AnnData subset to SingleCellAssay object...\")\n",
    "    \n",
    "    robjects.r('''\n",
    "    library(MAST)\n",
    "    FromMatrix <- function(counts, cData, fData) {\n",
    "        sca <- FromMatrix(counts, cData, fData)\n",
    "        return(sca)\n",
    "    }\n",
    "    ''')\n",
    "\n",
    "    counts_dense = adata_subset.X.todense()\n",
    "    cData = adata_subset.obs\n",
    "    fData = adata_subset.var\n",
    "    \n",
    "    # Convert data to the appropriate R data structures\n",
    "    r_counts = pandas2ri.py2rpy(pd.DataFrame(counts_dense))\n",
    "    r_cData = pandas2ri.py2rpy(cData)\n",
    "    r_fData = pandas2ri.py2rpy(fData)\n",
    "    \n",
    "    sca = robjects.r['FromMatrix'](r_counts, r_cData, r_fData)\n",
    "    return sca\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Steps followed in defining functions\n",
    "\n",
    "*Find_de_MAST*\n",
    "\n",
    "For performing differential gene expression analysis\n",
    "\n",
    "1. Define and fit the model by considering n_genes_per_cell and genotype and \n",
    "2. Perform likelihood-ratio test for the condition of interest\n",
    "3. Extract datatable from summary with log-fold changes and p-values\n",
    "4. Crea te a list to store results generated in next steps\n",
    "5. For each gene in a cellstate , extract log fold change using natural basic logarithm and p-value and FDR and store temporarily in alist with defined column and cellstate\n",
    "6. Convert the coefficients to log2 base\n",
    "7. False discovery rate calculation (Multiple testing correction)\n",
    "8. Call the R function 'find_de' for dsifferential expression analysis and run on sca object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to find differential expression using MAST\n",
    "def find_de_MAST(sca):\n",
    "    print(\"Finding differentially expressed genes...\")\n",
    "    robjects.r('''\n",
    "    library(MAST)\n",
    "    library(data.table)\n",
    "    \n",
    "    find_de <- function(sca) {\n",
    "    zlmCond <- zlm(~ n_genes_per_cell + cell_states + genotype, sca)\n",
    "    summaryCond <- summary(zlmCond, doLRT = TRUE)\n",
    "    summaryDt <- summaryCond$datatable\n",
    "    \n",
    "    results <- list()\n",
    "    \n",
    "    cell_states <- unique(summaryDt[component == 'H', contrast])\n",
    "    \n",
    "    for (cell_state in cell_states) {\n",
    "        if (cell_state != 'n_genes_per_cell'){\n",
    "            cell_state_lfc <- summaryDt[contrast == cell_state & component == 'logFC', .(primerid, coef)]\n",
    "            cell_state_p <- summaryDt[contrast == cell_state & component == 'H', .(primerid, `Pr(>Chisq)`)]\n",
    "            tmp <- merge(cell_state_lfc, cell_state_p, by='primerid')\n",
    "            tmp$log_fold_change <- tmp$coef / log(2)\n",
    "            tmp$FDR <- p.adjust(tmp$`Pr(>Chisq)`, 'fdr')\n",
    "            tmp$cell_states <- cell_state\n",
    "            \n",
    "            # Remove extra columns if any before renaming\n",
    "            tmp <- tmp[, .(primerid, log_fold_change, `Pr(>Chisq)`, FDR, cell_states)]\n",
    "                \n",
    "            colnames(tmp) <- c('gene_id', 'log_fold_change', 'p_value', 'FDR', 'cell_states')\n",
    "\n",
    "            \n",
    "            results[[cell_state]] <- tmp\n",
    "        }\n",
    "     }\n",
    "\n",
    "        results <- lapply(results, na.omit)\n",
    "        return(results)\n",
    "    }\n",
    "    ''')\n",
    "\n",
    "    print(\"Differentially expressed genes found.\")\n",
    "\n",
    "    result = robjects.r['find_de'](sca)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing genotype: Mdx, cell state: vCM2\n",
      "Converting AnnData subset to SingleCellAssay object...\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "# Loop through each genotype and each cell state\n",
    "genotypes = adata.obs['genotype'].unique()\n",
    "cell_states = adata.obs['cell_states'].unique()\n",
    "\n",
    "all_results = []\n",
    "\n",
    "for genotype in genotypes:\n",
    "    for cell_state in cell_states:\n",
    "        target = f\"{genotype}_{cell_state}\"\n",
    "        print(f\"Processing genotype: {genotype}, cell state: {cell_state}\")\n",
    "        \n",
    "        adata_subset = subset_data(adata, cell_state, genotype)\n",
    "        \n",
    "        if adata_subset.shape[0] == 0:\n",
    "            print(f\"No data found for genotype: {genotype}, cell state: {cell_state}\")\n",
    "            continue\n",
    "        \n",
    "        sca = convert_to_sca(adata_subset)\n",
    "        \n",
    "        de_results = find_de_MAST(sca)\n",
    "        \n",
    "        if cell_state in de_results:\n",
    "            de_results_df = pandas2ri.rpy2py_dataframe(de_results[cell_state])\n",
    "            \n",
    "            if not de_results_df.empty:\n",
    "                output_file = f'/home/skolla/Github/hofmann_dmd/hofmann_dmd/DE/de_results/DE_{genotype}_{cell_state}.txt'\n",
    "                de_results_df.to_csv(output_file, sep='\\t', index=False)\n",
    "                print(f\"Saved DE results for {genotype}_{cell_state} to {output_file}\")\n",
    "                all_results.append(de_results_df)\n",
    "            else:\n",
    "                print(f\"No differential expression results found for {genotype}_{cell_state}.\")\n",
    "        else:\n",
    "            print(f\"No results found for {genotype}_{cell_state}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Writing files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Steps followed \n",
    "\n",
    "1. Convert the results to a pandas DataFrame\n",
    "2. Savinng separate files for each cell state and a siungle file withh all results\n",
    "3. Adding MAST results to anndata and writing a new anndata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_results_df = {cell_state: pandas2ri.rpy2py_dataframe(df) for cell_state, df in de_results.items()}\n",
    "\n",
    "for cell_state, df in de_results_df.items():\n",
    "    file_name = f\"/home/skolla/Github/hofmann_dmd/hofmann_dmd/DE/de_results/DE_mdx_{cell_state}.txt\"\n",
    "    df.to_csv(file_name, sep='\\t', index=False)\n",
    "    print(f\"Saved DE results for {cell_state} to {file_name}\")\n",
    "all_results = pd.concat(de_results_df.values(), keys=de_results_df.keys())\n",
    "all_results.to_csv('/home/skolla/Github/hofmann_dmd/hofmann_dmd/DE/de_results/DE_mdx-all.txt', sep='\\t', index=False)\n",
    "\n",
    "adata.uns['MAST_results'] = de_results_df\n",
    "\n",
    "try:\n",
    "    sc_toolbox.tools.de_res_to_anndata(\n",
    "        adata,\n",
    "        all_results,\n",
    "        groupby=\"cell_states\",\n",
    "        gene_id_col='gene_id',\n",
    "        score_col='log_fold_change',\n",
    "        pval_col='p_value',\n",
    "        pval_adj_col=\"FDR\",\n",
    "        lfc_col='log_fold_change',\n",
    "        key_added='MAST_results'\n",
    "    )\n",
    "except ValueError as e:\n",
    "    print(f\"Error updating AnnData: {e}\")\n",
    "\n",
    "print(\"Updated AnnData object:\")\n",
    "print(adata)\n",
    "adata.write_h5ad('/home/skolla/Github/hofmann_dmd/hofmann_dmd/DE/de_results/DE_wt_adata.h5ad')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Steps followed \n",
    "\n",
    "1. Set thresholds\n",
    "2. Define a function\n",
    "3. Extracting MAST results into a dataframe and ensuring it has required columns\n",
    "4. Set index to gene names\n",
    "5. Filter based on thresholds and extract genes\n",
    "6. Create a output directory if doesnt exist\n",
    "7. Plot heatmap for each cellstate by subsetting to desired cellstate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FDR = 0.01\n",
    "LOG_FOLD_CHANGE = 1.5\n",
    "TOP_N_GENES = 50\n",
    "\n",
    "def plot_heatmap(adata, group_key, groupby=\"cell_states\", output_dir=\"plots\"):\n",
    "    if group_key not in adata.uns:\n",
    "        raise ValueError(f\"Group key '{group_key}' not found in adata.uns\")\n",
    "    \n",
    "    de_results_dict = adata.uns[group_key]\n",
    "    \n",
    "    # Check if de_results is a dictionary of DataFrames\n",
    "    if isinstance(de_results_dict, dict) and all(isinstance(v, pd.DataFrame) for v in de_results_dict.values()):\n",
    "        # Concatenate all DataFrames in the dictionary\n",
    "        res = pd.concat(de_results_dict.values(), keys=de_results_dict.keys(), names=['cell_states', 'index'])\n",
    "    else:\n",
    "        raise ValueError(f\"Unexpected format for differential expression results in adata.uns['{group_key}']\")\n",
    "    \n",
    "    required_columns = ['gene_id', 'log_fold_change', 'p_value', 'FDR']\n",
    "    for col in required_columns:\n",
    "        if col not in res.columns:\n",
    "            raise KeyError(f\"Expected column '{col}' not found in differential expression results.\")\n",
    "    \n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    \n",
    "    unique_groups = adata.obs[groupby].unique()\n",
    "\n",
    "    for group in unique_groups:\n",
    "        group_key_full = f'cell_states{group}'\n",
    "        \n",
    "        if group_key_full not in de_results_dict:\n",
    "            print(f\"No results found for group: {group}\")\n",
    "            continue\n",
    "\n",
    "        print(f\"Plotting heatmap for group: {group}\")\n",
    "        \n",
    "        # Filter results for the current group\n",
    "        group_res = de_results_dict[group_key_full]\n",
    "        \n",
    "        filtered_res = group_res[\n",
    "            (group_res[\"FDR\"] < FDR) & (abs(group_res[\"log_fold_change\"]) > LOG_FOLD_CHANGE)\n",
    "        ].sort_values(by=[\"log_fold_change\"], ascending=False)\n",
    "        \n",
    "        # Get top genes for the current cell state\n",
    "        top_genes = filtered_res.head(TOP_N_GENES)\n",
    "        \n",
    "        markers = list(top_genes['gene_id'].unique())\n",
    "\n",
    "        if len(markers) == 0:\n",
    "            print(f\"No significant genes found for group: {group}\")\n",
    "            continue\n",
    "        \n",
    "        adata_group = adata[adata.obs[groupby] == group].copy()\n",
    "        adata_group = adata_group[:, adata_group.var_names.isin(markers)]\n",
    "        \n",
    "        # Ensure that markers are displayed correctly\n",
    "        filename = f\"{group}_mdx_heatmap.png\"\n",
    "        filepath = os.path.join(output_dir, filename)\n",
    "\n",
    "        sc.pl.heatmap(\n",
    "            adata_group,\n",
    "            var_names=markers,\n",
    "            groupby=groupby,\n",
    "            swap_axes=True,\n",
    "            show=True,\n",
    "        )\n",
    "        \n",
    "        plt.title(f\"Heatmap - {group}_mdx\")\n",
    "        plt.ylabel(\"Gene Names\") \n",
    "        plt.savefig(filepath)\n",
    "        print(f\"Heatmap saved to {filepath}\")\n",
    "        plt.close()\n",
    "\n",
    "# Call the function\n",
    "plot_heatmap(adata, group_key=\"MAST_results\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FDR = 0.01\n",
    "LOG_FOLD_CHANGE = 1.0\n",
    "TOP_N_GENES = 50\n",
    "\n",
    "def plot_dotplot(adata, group_key, genotype_key, groupby=\"cell_states\", output_dir=\"figures\"):\n",
    "    if group_key not in adata.uns:\n",
    "        raise ValueError(f\"Group key '{group_key}' not found in adata.uns\")\n",
    "    \n",
    "    if genotype_key not in adata.obs:\n",
    "        raise ValueError(f\"Genotype key '{genotype_key}' not found in adata.obs\")\n",
    "    \n",
    "    de_results_dict = adata.uns[group_key]\n",
    "\n",
    "    if isinstance(de_results_dict, dict) and all(isinstance(v, pd.DataFrame) for v in de_results_dict.values()):\n",
    "        res = pd.concat(de_results_dict.values(), keys=de_results_dict.keys(), names=['cell_states', 'index'])\n",
    "    else:\n",
    "        raise ValueError(f\"Unexpected format for differential expression results in adata.uns['{group_key}']\")\n",
    "    \n",
    "    required_columns = ['gene_id', 'log_fold_change', 'p_value', 'FDR']\n",
    "    for col in required_columns:\n",
    "        if col not in res.columns:\n",
    "            raise KeyError(f\"Expected column '{col}' not found in differential expression results.\")\n",
    "    \n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir)\n",
    "    \n",
    "    unique_groups = adata.obs[groupby].unique()\n",
    "\n",
    "    for group in unique_groups:\n",
    "        group_key_full = f'cell_states{group}'\n",
    "        \n",
    "        if group_key_full not in de_results_dict:\n",
    "            print(f\"No results found for group: {group}\")\n",
    "            continue\n",
    "\n",
    "        print(f\"Plotting dot plot for group: {group}\")\n",
    "        \n",
    "        group_res = de_results_dict[group_key_full]\n",
    "        \n",
    "        filtered_res = group_res[\n",
    "            (group_res[\"FDR\"] < FDR) & (abs(group_res[\"log_fold_change\"]) > LOG_FOLD_CHANGE)\n",
    "        ].sort_values(by=[\"log_fold_change\"], ascending=False)\n",
    "        \n",
    "        top_genes = filtered_res.head(TOP_N_GENES)\n",
    "        markers = list(top_genes['gene_id'].unique())\n",
    "        if len(markers) == 0:\n",
    "            print(f\"No significant genes found for group: {group}\")\n",
    "            continue\n",
    "        \n",
    "        adata_group = adata[adata.obs[groupby] == group].copy()\n",
    "        \n",
    "        genotypes = adata_group.obs[genotype_key].unique()\n",
    "        if len(genotypes) > 0:\n",
    "            genotype = genotypes[0]\n",
    "            title = f\"Dot Plot for {group} - Genotype: {genotype}\"\n",
    "        else:\n",
    "            title = f\"Dot Plot for {group}\"\n",
    "        \n",
    "        filename = f\"{group}_dotplot.png\"\n",
    "        filepath = os.path.join(output_dir, filename)\n",
    "\n",
    "        sc.pl.dotplot(\n",
    "            adata_group,\n",
    "            var_names=markers,\n",
    "            groupby=groupby,\n",
    "            cmap=\"RdYlBu_r\",\n",
    "            title=title,\n",
    "            show=True,\n",
    "            save=filepath\n",
    "        )\n",
    "        \n",
    "        print(f\"Dot plot saved to {filepath}\")\n",
    "        plt.close()\n",
    "\n",
    "\n",
    "plot_dotplot(adata, group_key=\"MAST_results\", genotype_key=\"genotype\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 8257 × 16060\n",
       "    obs: 'cell_source', 'cell_type', 'donor', 'cell_states', 'genotype', 'compartment', 'object', 'samples', 'n_counts', 'batch'\n",
       "    var: 'gene_ids-CMC'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read_h5ad('/home/skolla/Github/hofmann_dmd/hofmann_dmd/data/concatenated_CMC-16_07_2024.h5ad')\n",
    "\n",
    "\n",
    "desired_states = ['vCM1', 'vCM2', 'vCM3', 'vCM4']\n",
    "filtered_adata = adata[adata.obs['cell_states'].isin(desired_states)]\n",
    "\n",
    "filtered_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from upsetplot import UpSet\n",
    "import matplotlib.pyplot as plti"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "The DataFrame has values in its index that are not boolean",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_383654/3971429756.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[0;31m# Create the UpSet plot\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 30\u001b[0;31m \u001b[0mupset\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mUpSet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcombinations\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     31\u001b[0m \u001b[0mupset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtitle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Upset Plot of Gene Counts by Genotype and Cell States'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniforge3/envs/mast_env/lib/python3.7/site-packages/upsetplot/plotting.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, orientation, sort_by, sort_categories_by, subset_size, sum_over, min_subset_size, max_subset_size, max_subset_rank, min_degree, max_degree, facecolor, other_dots_color, shading_color, with_lines, element_size, intersection_plot_elements, totals_plot_elements, show_counts, show_percentages, include_empty_subsets)\u001b[0m\n\u001b[1;32m    353\u001b[0m             \u001b[0mmax_degree\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmax_degree\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    354\u001b[0m             \u001b[0mreverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_horizontal\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 355\u001b[0;31m             \u001b[0minclude_empty_subsets\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minclude_empty_subsets\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    356\u001b[0m         )\n\u001b[1;32m    357\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcategory_styles\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniforge3/envs/mast_env/lib/python3.7/site-packages/upsetplot/plotting.py\u001b[0m in \u001b[0;36m_process_data\u001b[0;34m(df, sort_by, sort_categories_by, subset_size, sum_over, min_subset_size, max_subset_size, max_subset_rank, min_degree, max_degree, reverse, include_empty_subsets)\u001b[0m\n\u001b[1;32m     46\u001b[0m         \u001b[0mmin_degree\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmin_degree\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m         \u001b[0mmax_degree\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmax_degree\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 48\u001b[0;31m         \u001b[0minclude_empty_subsets\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minclude_empty_subsets\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     49\u001b[0m     )\n\u001b[1;32m     50\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniforge3/envs/mast_env/lib/python3.7/site-packages/upsetplot/reformat.py\u001b[0m in \u001b[0;36mquery\u001b[0;34m(data, present, absent, min_subset_size, max_subset_size, max_subset_rank, min_degree, max_degree, sort_by, sort_categories_by, subset_size, sum_over, include_empty_subsets)\u001b[0m\n\u001b[1;32m    367\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    368\u001b[0m     \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0magg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_aggregate_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubset_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msum_over\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 369\u001b[0;31m     \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_check_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    370\u001b[0m     \u001b[0mgrand_total\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0magg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    371\u001b[0m     category_totals = [\n",
      "\u001b[0;32m~/miniforge3/envs/mast_env/lib/python3.7/site-packages/upsetplot/reformat.py\u001b[0m in \u001b[0;36m_check_index\u001b[0;34m(df)\u001b[0m\n\u001b[1;32m     71\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m}\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mlevel\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlevels\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m         raise ValueError(\n\u001b[0;32m---> 73\u001b[0;31m             \u001b[0;34m\"The DataFrame has values in its index that are not \"\u001b[0m \u001b[0;34m\"boolean\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     74\u001b[0m         )\n\u001b[1;32m     75\u001b[0m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: The DataFrame has values in its index that are not boolean"
     ]
    }
   ],
   "source": [
    "# Extract gene counts for a specific gene (adjust index as needed)\n",
    "gene_of_interest = 0  # Adjust this index or use gene names if available\n",
    "if hasattr(filtered_adata.X, 'toarray'):\n",
    "    gene_counts = filtered_adata.X[:, gene_of_interest].toarray().flatten()\n",
    "else:\n",
    "    gene_counts = filtered_adata.X[:, gene_of_interest].flatten()\n",
    "\n",
    "# Create DataFrame\n",
    "df = pd.DataFrame({\n",
    "    'GeneCounts': gene_counts,\n",
    "    'Genotype': filtered_adata.obs['genotype'].values,\n",
    "    'CellStates': filtered_adata.obs['cell_states'].values\n",
    "})\n",
    "\n",
    "# Create a DataFrame where each row represents the presence of a combination\n",
    "# Using cross-tabulation to represent combinations\n",
    "combination_df = pd.crosstab(index=df['Genotype'], columns=df['CellStates'])\n",
    "\n",
    "# Create a MultiIndex with all combinations of Genotype and CellStates\n",
    "all_combinations = pd.MultiIndex.from_product([combination_df.index, combination_df.columns], names=['Genotype', 'CellStates'])\n",
    "\n",
    "# Reindex the combination_df to include all possible combinations\n",
    "combination_df = combination_df.reindex(all_combinations, fill_value=0)\n",
    "\n",
    "# Convert this DataFrame into a Series with multi-index for UpSet\n",
    "combinations = combination_df.stack()\n",
    "combinations.name = 'count'\n",
    "\n",
    "# Create the UpSet plot\n",
    "upset = UpSet(combinations)\n",
    "upset.plot()\n",
    "plt.title('Upset Plot of Gene Counts by Genotype and Cell States')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mast_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
